File: go/leetcode/insert_interval.go
package leetcode

func insert(intervals [][]int, newInterval []int) [][]int {
	if newInterval[1] < intervals[0][0] {
		return append([][]int{newInterval}, intervals...)
	}

	if newInterval[0] > intervals[len(intervals)-1][1] {
		return append(intervals, newInterval)
	}

	low := 0
	high := len(intervals) - 1
	var mid int

	for low <= high {
		mid = low + ((high - low) / 2)

		if newInterval[0] > intervals[mid][0] {
			high = mid - 1
		} else if newInterval[0] < intervals[mid][0] {
			low = mid + 1
		} else {
			break
		}
	}
}

File: cpp/InsertInterval/InsertInterval.cpp
#include<iostream>
#include<vector>
#include <utility>

using std::vector;
using std::min;
using std::max;

class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {

        if (intervals.empty())
            return vector<vector<int>>{newInterval};

        if(newInterval.at(0) > intervals.at(intervals.size()-1).at(1))
        {
            intervals.push_back(newInterval);
            return intervals;
        }

        int i = 0;    
        vector<vector<int>> result;
        while (i < intervals.size() && newInterval.at(0) > intervals.at(i).at(1))
            i++;
        
        result.insert(result.end(), intervals.begin(), intervals.begin()+i);

        while (i < intervals.size() && newInterval.at(1) >= intervals.at(i).at(0))
        {
            newInterval = {
                min(newInterval.at(0), intervals.at(i).at(0)),
                max(newInterval.at(1), intervals.at(i).at(1))
            };
            i++;
        }
        result.push_back(newInterval);
        result.insert(result.end(), intervals.begin()+i, intervals.end());
        
        return result;
    }
};